### DISCUSSION TITLE
Using Reach Stdlib to make a single transaction for authentication

### DISCUSSION THREAD

**Discussion by cooleydw494** [reach-sh/reach-lang#1373] [General]
I'm using Vue, JACK-COM/reachduck, and a php/laravel back-end API. I'm also using a custom authentication setup for the API with the intention of generating a challenge (a nonce attached to a user record) and having the user sign a transaction with it on the front-end, and then sending back the base64 encoded txn data to be verified. I didn't write the verification code, and it is possible the issue is there, but I'm posting this because I'm not sure if what I'm seeing on the Reach stdlib side is correct, or if I'm understanding it correctly.

Here is my wallet connect implementation, which uses reachduck to connect a wallet via various methods like MyAlgo. I then request a challenge from my API, use setSigningMonitor to collect txn data, and have the user authorize a 0-amount transfer with the challenge as the note. I am using a method called str2arr to convert the challenge to a UInt8 array (this may be what I'm doing wrong, I'm not sure exactly how this is handled, or if it is compatible with what I'm trying to accomplish, I just know that's the format of the note field). Then I wait for the txn to populate via setSigningMonitor, and I access the base64 encoded txn data from the evt object and send it back to my API for verification. I'm not posting the back-end implementation, because if that's where the problem lies, it isn't really fair to ask for help here. My intention is to make sure I'm understanding what's happening on the Reach stdlib side of things correctly. In any case, all appears to work, but authentication fails.

These are the Vue methods I'm using to get the job done:

    `str2arr(str) {
      const utf8 = decodeURI(encodeURIComponent(str))
      const array = new Uint8Array(utf8.split('').map(function (item) {
        return item.charCodeAt(0)
      }))
      console.log(array);
      return array;
    },
    async connectTo(provider) {
      this.showWalletSelect = false
      this.connecting = true
      await connectWallet(provider).then(() => {
        console.log('Wallet connected with provider: ' + provider)
        get(`auth/request-challenge/${this.store.address}`).then(async res => {
          console.log(res)
          try {
            const reach = loadStdlib()
            reach.setSigningMonitor(async (evt, pre, post) =>
                this.txns.push({evt, pre: await pre, post: await post}));

            await reach.transfer(this.store.account, this.store.account, 0, null, {note: this.str2arr(res.challenge)})
            do {
              await this.sleep(250)
            } while (this.txns.length === 0)
            const signedTxn = this.txns.pop().evt[0].txn
            post('auth/login', {
              algorand_address: this.store.address, signed_tx: signedTxn
            }).then(res => console.log(res))
          } catch (err) {
            console.log('Failed to sign authentication challenge transaction or login', err)
          }
        })
      }).catch(err => {
        alert(err)
      })
      this.connecting = false
    },`
    
    
This is a challenge from my API: **Oawn6iIEPJmpYvBj**
    
This is the Uint8Array I convert it to and pass as the note:
`Uint8Array(16) [79, 97, 119, 110, 54, 105, 73, 69, 80, 74, 109, 112, 89, 118, 66, 106, buffer: ArrayBuffer(16), byteLength: 16, byteOffset: 0, length: 16, Symbol(Symbol.toStringTag): 'Uint8Array']`

This is the signed txn data I send back: `iaNmZWXNA+iiZnbOAWm7xKNnZW6sdGVzdG5ldC12MS4womdoxCBIY7UYpLPITsgQ8i1PEIHLD3HwWaesIN7GL39w5Qk6IqJsds4Bab+spG5vdGXEEE9hd242aUlFUEptcFl2QmqjcmN2xCCWc0HZN5pnwVAh0Kw7GnPGl7pAiKofiUZrYnF0i881maNzbmTEIJZzQdk3mmfBUCHQrDsac8aXukCIqh+JRmticXSLzzWZpHR5cGWjcGF5

This is that txn data run through atob: `£feeÍ\u0003è¢fvÎ\u0001i»Ä£gen¬testnet-v1.0¢ghÄ Hcµ\u0018¤³ÈNÈ\u0010ò-O\u0010Ë\u000fqðY§¬ ÞÆ/på\t:\"¢lvÎ\u0001i¿¬¤noteÄ\u0010Oawn6iIEPJmpYvBj£rcvÄ sAÙ7gÁP!Ð¬;\u001asÆº@ª\u001fFkbqtÏ5£sndÄ sAÙ7gÁP!Ð¬;\u001asÆº@ª\u001fFkbqtÏ5¤type£pay

So in summation, I thought I managed to find a way to accomplish what I needed for back-end API authentication using the connection I've already established via Reach for everything else I'm doing. It isn't working however, and while there could be other reasons, I want to verify that I'm not misunderstanding something about the front-end implementation with stdlib.

Thanks in advance for any commentary.

**Comment by jeapostrophe** [comment]
I think I am stupid. I don't understand what the question is. You say "I thought I managed to find a way [... but ...] It isn't working". I don't actually understand what you're trying to do. I don't know what "reachduck" is either, FWIW.

**Reply by cooleydw494** [reply]
Oh boy haha, sorry Jay. It definitely isn't you, I was afraid it wouldn't be clear. For the record, reachduck is just a tool for using the Reach stdlib and is how I'm connecting the wallet initially. I don't think it is relevant.

Really, the problem occurs when I try to validate the signed transaction on the back-end, and my assumption is that the Reach stdlib is working correctly. I posted all of this (a confusing mix of too much/too little context perhaps) to make sure I wasn't doing something wrong on the Reach side before I assume the problem is elsewhere.

**My intention is to use the Reach stdlib to have the user sign a 0 payment transaction with a note and then access the base64 encoded signed txn data to send back to my API. I tried to boil this down to clear questions below. If it seems the answers are obvious, it is because I'm trying to double-check what already seems to be correct.**

Is using the Reach stdlib to create this transaction, and storing the txn data with setSigningMonitor, and sending back the evt[0].txn giving me a base64 encoded string of the transaction data (what I need)?

Is passing a string which is converted to UInt8 Array as the note field via the Reach stdlib the correct way to use it?

If those two things hold up, is there anything about the way that Reach handles things that would differ from any other method of doing this?


Again, apologies for the confusion, if your reaction to this is that I'm doing something with Reach that makes sense / is working, and you don't get what else is going on, I think that validates my understanding :)

**Reply by cooleydw494** [reply]
I have figured it out. The things I was using Reach for worked correctly, but the txn data just didn't include everything I needed. I queried the transaction using the indexer and used the extra data to achieve what I need.

Apologies for the lengthy discussion topic. Thanks for taking a look!

**Reply by cooleydw494** [reply]
If you are interested however, I'm up to discussing the value stored in evt[0]. It wasn't as I was expecting. Actually, the data within 'txn' was what I expected, but the package I was using to implement auth was expecting alongside 'txn' to see 'sgnr', 'sig', maybe something else I'm not remembering, and I believe it was expecting all of that to be encoded together, though that isn't necessary. The details are a bit beyond my own understanding. Tomas, the package author told me that the format expected is what is returned from the Java/JS sdks when a payment transaction is signed.

Don't get me wrong, I certainly don't see it as a problem with Reach and I know you aren't following the package I'm using, just curious/interested in how the txn data is exposed, because if I'm not mistaken its coming from algosdk. I certainly wouldn't complain if I could use the txn data exposed by Reach without cobbling things together, but I have a working solution now!

Thanks, and apologies for the mess!

**Reply by jeapostrophe** [reply]
I think I understand now. In the Algorand world, there are many different ways of representing transactions. You can have a signed byte string, you have an encoded byte string (they use msgpack), you can decode the byte string and get the protocol labels (see the "codec" column here --- https://developer.algorand.org/docs/get-details/transactions/transactions/#payment-transaction ), or you can have an object with "human readable" keys (like `amount` rather than `amt`) as in the JS lib --- https://algorand.github.io/js-algorand-sdk/modules.html#makePaymentTxnWithSuggestedParamsFromObject
